File: scripts/lib/healthchecks.py
"""Python mirror of supabase/functions/_shared/healthchecks.ts.

Telegram was removed on 2026-08-25 (docs/cardledger-build-spec.md §10
AMENDMENT). healthchecks.io is now the only out-of-band alarm across BOTH
runtimes: Edge Functions (Deno, `_shared/healthchecks.ts`) and this GitHub
Actions / `scripts/` runtime (Python). `scripts/` cannot import the Deno
module, so this file duplicates its API on purpose rather than being a
second, drifted mechanism — see that file for the fuller rationale on why
healthchecks.io specifically (a dead-man's-switch: it alerts when pings
*stop*, which is what catches a paused project or a silently dropped cron
schedule).

Three request shapes, same base URL:
    POST <url>       -> success ping ("still alive")
    POST <url>/fail   -> explicit failure, triggers an alert email immediately
    POST <url>/log    -> informational, recorded but does not alert

Callers choose which one: reserve /fail for genuine system-health problems
(the reconcile run itself crashing) and use /log for data-quality signals
(an above-threshold miss rate) — using /fail for routine data-quality noise
trains the operator to ignore the alarm, which defeats the point of having
one.

Same requirement the old telegram helper (and its Deno replacement) call
out: the one function responsible for surfacing every failure mode must not
itself fail silently. So: a short timeout, this never raises, and every
call returns a result the caller can check instead of the failure being
swallowed. A missing HEALTHCHECKS_PING_URL logs loudly and returns
ok=False rather than raising — a missing alarm URL must not crash the
reconciliation run that was trying to report a *different* problem.
"""
from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from typing import Literal

import requests

PING_TIMEOUT_SECONDS = 5


@dataclass(frozen=True)
class HealthcheckPingResult:
    ok: bool
    # Populated when ok is False. Safe to log.
    error: str | None = None


def _ping(suffix: Literal["", "/fail", "/log"], body: str | None = None) -> HealthcheckPingResult:
    base_url = os.environ.get("HEALTHCHECKS_PING_URL")
    if not base_url:
        error = "HEALTHCHECKS_PING_URL not configured"
        print(f"healthchecks.io ping skipped ({suffix or 'success'}): {error}", file=sys.stderr)
        return HealthcheckPingResult(ok=False, error=error)

    try:
        if body is not None:
            res = requests.post(f"{base_url}{suffix}", data=body.encode("utf-8"), timeout=PING_TIMEOUT_SECONDS)
        else:
            res = requests.get(f"{base_url}{suffix}", timeout=PING_TIMEOUT_SECONDS)
        if not res.ok:
            error = f"healthchecks.io returned HTTP {res.status_code}"
            print(f"healthchecks.io ping failed ({suffix or 'success'}): {error}", file=sys.stderr)
            return HealthcheckPingResult(ok=False, error=error)
        return HealthcheckPingResult(ok=True)
    except requests.RequestException as exc:
        error = str(exc)
        print(f"healthchecks.io ping failed ({suffix or 'success'}): {error}", file=sys.stderr)
        return HealthcheckPingResult(ok=False, error=error)


def ping_success() -> HealthcheckPingResult:
    """Success ping — "still alive". No body: healthchecks needs nothing more."""
    return _ping("")


def ping_fail(reason: str) -> HealthcheckPingResult:
    """Failure ping — triggers an alert email immediately. Reserve for
    genuine system-health problems: the run itself crashing, a source gone
    silent. ``reason`` becomes the ping body, so keep it short and specific
    — it's what shows up in the alert.
    """
    return _ping("/fail", reason)


def ping_log(detail: str) -> HealthcheckPingResult:
    """Informational ping — recorded on the healthchecks.io dashboard but
    does NOT alert. For data-quality issues (e.g. an above-threshold miss
    rate) that are not themselves evidence the system is down.
    """
    return _ping("/log", detail)

File: supabase/functions/_shared/healthchecks.ts
// The only out-of-band alarm left once Telegram is gone (operator decision,
// 2026-08-25, see docs/architecture.md §2), matching the `heartbeat` job's
// current description in §7: Supabase Cron has no failure alerting of its
// own, and the per-source silence check exists because a bank quietly
// reverting an alert threshold is otherwise invisible to an aggregate
// "did anything ingest today" check. healthchecks.io is a dead-man's-
// switch: it alerts by email when pings *stop*, which is what catches a
// paused project or a silently
// dropped cron schedule that no in-process check could ever see.
//
// Three request shapes, same base URL:
//   POST <url>        -> success ping ("still alive")
//   POST <url>/fail    -> explicit failure, triggers an alert email immediately
//   POST <url>/log     -> informational, recorded but does not alert
//
// Callers choose which one: reserve /fail for genuine system-health
// problems (per-source silence, a held watermark) and use /log for
// data-quality issues (parse-failure spikes, permanently-unparseable
// messages) — using /fail for routine data-quality noise trains the
// operator to ignore the alarm, which defeats the point of having one.
//
// Same requirement the old _shared/telegram.ts called out: "the one
// function responsible for surfacing every failure mode can itself fail
// silently" must not happen here either. So: a short timeout, this never
// throws, and every call returns a result the caller can check instead of
// swallowing it. A missing HEALTHCHECKS_PING_URL logs loudly and returns
// ok: false rather than throwing — a missing alarm URL must not crash the
// function that was trying to report a *different* problem.

const PING_TIMEOUT_MS = 5_000;

export interface HealthcheckPingResult {
  ok: boolean;
  /** Populated when ok === false. Safe to log. */
  error?: string;
}

async function ping(suffix: "" | "/fail" | "/log", body?: string): Promise<HealthcheckPingResult> {
  const baseUrl = Deno.env.get("HEALTHCHECKS_PING_URL");
  if (!baseUrl) {
    const error = "HEALTHCHECKS_PING_URL not configured";
    console.error(`healthchecks.io ping skipped (${suffix || "success"}): ${error}`);
    return { ok: false, error };
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), PING_TIMEOUT_MS);
  try {
    const res = await fetch(`${baseUrl}${suffix}`, {
      // POST (rather than GET) whenever there's a body: healthchecks.io
      // stores the POST body (up to 10KB) and shows it in the ping
      // details / alert email. The plain success ping carries nothing
      // worth recording, so it stays a GET, matching the original
      // heartbeat behaviour exactly.
      method: body !== undefined ? "POST" : "GET",
      body,
      signal: controller.signal,
    });
    await res.text().catch(() => {}); // drain so the connection can be released
    if (!res.ok) {
      const error = `healthchecks.io returned HTTP ${res.status}`;
      console.error(`healthchecks.io ping failed (${suffix || "success"}): ${error}`);
      return { ok: false, error };
    }
    return { ok: true };
  } catch (err) {
    const error = err instanceof Error ? err.message : String(err);
    console.error(`healthchecks.io ping failed (${suffix || "success"}): ${error}`);
    return { ok: false, error };
  } finally {
    clearTimeout(timeout);
  }
}

/** Success ping — "still alive". No body: healthchecks needs nothing more. */
export function pingSuccess(): Promise<HealthcheckPingResult> {
  return ping("");
}

/**
 * Failure ping — triggers an alert email immediately. Reserve for genuine
 * system-health problems: a source gone silent, a watermark that is being
 * held and will retry forever until resolved. `reason` becomes the ping
 * body, so keep it short and specific — it's what shows up in the alert.
 */
export function pingFail(reason: string): Promise<HealthcheckPingResult> {
  return ping("/fail", reason);
}

/**
 * Informational ping — recorded on the healthchecks.io dashboard but does
 * NOT alert. For data-quality issues (a parse failure, an unparseable
 * message) that are not themselves evidence the system is down.
 */
export function pingLog(detail: string): Promise<HealthcheckPingResult> {
  return ping("/log", detail);
}
